### ISSUE TITLE
Artifacts are installed with wrong permissions

### ISSUE THREAD

**Issue by maleadt** [JuliaLang/Pkg.jl#3269]
I'm having an issue with a rootfs artifact (i.e. containing a Linux OS) being extracted with the wrong permissions. The image in question is used in PkgEval.jl, and the problematic file is a lock whose `u+w` permissions are dropped.

First of all, here's the tarball and the permissions the lock file should have:

```
❯ wget https://github.com/JuliaCI/PkgEval.jl/releases/download/v0.1/debian-bullseye-20220818.tar.xz
❯ tar -tvf debian-bullseye-20220818.tar.xz | grep var/lib/dpkg/lock-frontend
-rw-r-----  0 tim    tim         0 Aug 18 16:19 ./var/lib/dpkg/lock-frontend
```

If I download this tarball using Pkg's `download_verify_unpack`, all seems well (an `o+r` permission is added, which isn't great, but also not terrible):

```
julia> Pkg.download_verify_unpack("https://github.com/JuliaCI/PkgEval.jl/releases/download/v0.1/debian-bullseye-20220818.tar.xz", nothing, mktempdir(); verbose=true, ignore_existence=true)
[ Info: Unpacking /var/folders/q0/wbdz2dpd0q9gtjk3wkj1fcnh0000gn/T/jl_ti7eJ2RSRC-download.xz into /var/folders/q0/wbdz2dpd0q9gtjk3wkj1fcnh0000gn/T/jl_EajM2J...
true

shell> ls -la /var/folders/q0/wbdz2dpd0q9gtjk3wkj1fcnh0000gn/T/jl_EajM2J/var/lib/dpkg/lock-frontend
-rw-r--r--  1 tim  staff  0 Dec  2 10:07 /var/folders/q0/wbdz2dpd0q9gtjk3wkj1fcnh0000gn/T/jl_EajM2J/var/lib/dpkg/lock-frontend
```

However, if I pack this up in an `Artifacts.toml`, I get the following:

```
julia> ArtifactUtils.add_artifact!("Artifacts.toml", "debian", "https://github.com/JuliaCI/PkgEval.jl/releases/download/v0.1/debian-bullseye-20220818.tar.xz"; lazy=true)

❯ cat Artifacts.toml
[debian]
git-tree-sha1 = "e5d2359bc94f9d9ddc771c2274c83324ca8cbbcc"
lazy = true
    [[debian.download]]
    sha256 = "072392d8635847de70665af512551038b04baca99b38e271d0eef1ea36d2833a"
    url = "https://github.com/JuliaCI/PkgEval.jl/releases/download/v0.1/debian-bullseye-20220818.tar.xz"

julia> using LazyArtifacts

julia> artifact"debian"
"/Users/tim/.julia/artifacts/e5d2359bc94f9d9ddc771c2274c83324ca8cbbcc"

shell> ls -la /Users/tim/.julia/artifacts/e5d2359bc94f9d9ddc771c2274c83324ca8cbbcc/var/lib/dpkg/lock-frontend
-r--r--r--  1 tim  staff  0 Dec  2 10:08 /Users/tim/.julia/artifacts/e5d2359bc94f9d9ddc771c2274c83324ca8cbbcc/var/lib/dpkg/lock-frontend
```

... and our `u+w` permission is gone. That's bad, here breaking `dpkg` in the containers spawned using this rootfs because it can't take the lock.

Tested on 1.8.2.

**Comment by giordano** [comment]
Is that because Tar.jl normalises permissions?

**Comment by maleadt** [comment]
I don't think so, because in https://github.com/JuliaCI/PkgEval.jl/pull/189 I'm not using Tar.jl directly to unpack this tarball, and permissions are fine.

**Comment by maleadt** [comment]
> I think the reason the areas of the ecosystem that I administer don't run into this is that I'm usually running as `root` on the inside of containers

Makes sense, but for PkgEval I don't want to run as root in order to spot packages doing questionable things (e.g., using `su` or `sudo`).

**Comment by KristofferC** [comment]
What could be done here? Stop setting the artifacts to read only?

**Comment by maleadt** [comment]
> Stop setting the artifacts to read only?

I would argue that's the expected behavior. But root filesystems are a fairly special case, and for most packages it's probably more interesting to have everything be read-only in order to detect accidental writes to the artifact store.

What about an additional tag in the `Artifacts.toml`? A default `chmod = true` for the current behavior, and `chmod = false` to support disabling it?

**Comment by KristofferC** [comment]
Should that choice be made at the `Artifact.toml` level or at the point when things are installed?

**Comment by maleadt** [comment]
In my case at least, it's mostly a property of the artifact. Maybe phrasing it as `readonly = true` ties it more to the artifact?

**Comment by staticfloat** [comment]
Because we identify artifacts by content-hash, it's a little risky to allow them to be made non-readonly, as we then break the implicit contract that this artifact can be shared by multiple environments/packages with no fear of them being "incorrect" in some way.

In practice, I'll just remind Tim that he broke the contract by setting the flag, and he has only himself to blame.  This should be a pretty easy thing to implement, and I agree this should be in `Artifacts.toml` on a per-artifact basis.

**Comment by staticfloat** [comment]
I think the name of the setting should be something like `set_read_only = false`.

**Comment by maleadt** [comment]
Yeah, I guess agree that this kinda violates the idea of the artifact store, and maybe we should simply not use it to store (extracted) root filesystems, if anything because there's additional issues like the fact that empty directories (as required by some tools) are not always dealt with correctly. I'd be fine with that as well.

**Comment by staticfloat** [comment]
I guess it comes down to convenience. If you are okay with using something that just downloads to a scratch space and using that, that’s easiest for the Pkg devs.

**Comment by maleadt** [comment]
Maybe one convenience feature then would be to support artifacts that are not to be extracted. Currently, I hand-roll the TOML parsing, while something like `extract = false` could allow me to simply look up the tarball.

**Comment by staticfloat** [comment]
So we’d have artifacts, lazy artifacts, and “utterly devoted to the sin of sloth” artifacts. :)
